_posts/2020-08-29-gsoc-webgpu.md

```diff
@@ -0,0 +1,98 @@
+- The contents of the rendered texture are copied to a buffer.
+- Buffer is mapped asynchronously for read.
+- The data read from the buffer is copied to a staging area in [`PresentionData`](https://github.com/servo/servo/blob/669b16f2c054bd038b7a3c69985076607e140b7f/components/webgpu/lib.rs#L1353-L1365). `PresentationData` stores the data and all the required machinery for this async presentation belt.
+- When WebRender wants to read the data, it locks on the data to prevent it from being altered during read. Data is served in the form of raw bytes.
+
+The above process is not the best one, but the only option available to us for now. This also causes a few empty frames to be rendered at the start.
+A good thing, though, is that this works on all platforms and is a great fallback path while we'll be adding hardware accelerate presentation in the future.
+
+## Buffer Mapping
+When the user issues an async buffer map operation, the operation is queued on the server-side and all devices polled at a regular interval of 100ms for the same. As soon as the map operation is complete, data is read and sent to the content process where it is stored in the Heap. The user can read and edit this data by accessing it's subranges via `GPUBuffer.getMappedRange()` which returns `ExternalArrayBuffer` pointing to the data in the Heap. On unmap, all the `ExternalArrayBuffer`s are detached, and if the buffer was mapped for write, data sent back to server for write to the actual resource.
+
+## Error Reporting
+To achieve maximum efficiency, WebGPU supports an asynchronous error model. The implementation keeps a stack of `ErrorScope`s that are responsible for capturing the errors that occur during operations performed in their scope. The user is responsible for pushing and popping an `ErrorScope` in the stack. Popping an `ErrorScope` returns a promise that is resolved to null if all the operations were successfull, otherwise it resolves to the first error that occurred.
+
+When an operation is issued, `scope_id` of the `ErrorScope` on the top of the stack is sent to the server with it and operation-count of the scope is incremented. The result of the operation can be described by the enum-
+
+	pub enum WebGPUOpResult {
```

Comment:
does Servo block support syntax-specific highlighting? In the blogs I seen you could use the regular "```rust" thing

Reply:
"```" wasn't working for me for code blocks. I'll quickly check again though.

Reply:
It worked. I must be missing something earlier.